### ISSUE TITLE
feature change proposal re melt.data.table and key'ing behavior

### ISSUE THREAD

**Issue by pearsonca** [Rdatatable/data.table#5490]
Per contributor guidelines, I'm outlining a proposed change concerning keyed `data.table`s and `melt.data.table` to get feedback before embarking on a PR. Briefly:

1. I'd prefer the default behavior to be roughly `melt.data.table(data, id.vars = key(data), ...)`, and
2. I'd like the return value of `melt.data.table` to be key'd roughly by `c(id.vars, variable.name)`

The cons I can think of are:
 - this changes behavior, and I can imagine it breaking some existing code that relies on the `id.vars = NULL` behavior for a key'd input and that behavior differs from what would happen assuming the key.
 - also possible, but less plausible, would be breaking behavior based on assumed (dis)order of return value
 - for some melts, key'ing the result could be an unnecessary performance hit

The pros amount to roughly this is the behavior I want 99% of the times I use `melt.data.table` and I'd like to type less (and troubleshoot fewer errors when I forget to do this). I haven't looked under the hood at `Cfmelt`, but it's plausible that could be more performant when melting by keys (or subset of keys) and that it might make sense to add a conditional for that (newly-default-so-now-more-likely) situation.

The cons can be somewhat mitigated, and the behavior changes indicate holding this sort of change for a breaking version. The breaking behavior can be addressed by minor dependant code changes (explicitly setting `id.vars = NULL`, setting a `key.return` or somesuch argument that can be set to `FALSE` or perhaps some logic that checks if the input is keyed, if that key is used for id vars, and if variable is a factor). The `keyed.return` argument would also address potential performance-minded desire for non-ordered outputs.

What considerations am I missing? I flipped though old issues searching on `melt` and `key` or `id.vars`, and didn't see any titles concerning this idea, but may have missed something or some other developer discussion of this kind of approach with reasons for rejecting it.

**Comment by jangorecki** [comment]
I am not really melt user, but if there is a chance for performance regression (or an extra overhead due to keying) then we need the worst case scenario benchmark to evaluate to costs.
If behavior you are requesting was not being raised before, and there won't be many +1 here, then I don't think we have good reasons to change the default.

**Comment by pearsonca** [comment]
Yep, agree re performance regression being the priority consideration. I think it's a bit complicated by this-method-only versus overall use (e.g. melt-then-setkey), but can imagine some reasonable benchmarks.

Also comfortable with the idea that my pet issue might not interest other users!

**Comment by pearsonca** [comment]
A colleague highlighted a different point regarding piping which made me realize that part of my annoyance here is having to write

```
{ ... something yielding a data.table ... } |> (\(dt) melt(dt, id.vars = key(dt), ...))
```
...because I don't actually have a `dt` around, when I want to write just a direct pipe into `melt`.

Allowing `id.vars` to be a function (applied to `data` argument) would deal with a lot of that annoyance.

**Comment by jangorecki** [comment]
For now the easiest you can do is just to make melt2 wrapper that changes the defaults.

**Comment by tdhock** [comment]
There are two separate proposals here. To make it concrete, consider iris
```r
> (two.iris = data.table(datasets::iris)[c(1,150)])
   Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
          <num>       <num>        <num>       <num>    <fctr>
1:          5.1         3.5          1.4         0.2    setosa
2:          5.9         3.0          5.1         1.8 virginica
```

First, default with no key is measure.vars=all numeric vars:
```r
> melt(two.iris)
     Species     variable value
      <fctr>       <fctr> <num>
1:    setosa Sepal.Length   5.1
2: virginica Sepal.Length   5.9
3:    setosa  Sepal.Width   3.5
4: virginica  Sepal.Width   3.0
5:    setosa Petal.Length   1.4
6: virginica Petal.Length   5.1
7:    setosa  Petal.Width   0.2
8: virginica  Petal.Width   1.8
Warning message:
In melt.data.table(two.iris) :
  id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [Species, ...]. Consider providing at least one of 'id' or 'measure' vars in future.
```
After setting key, result is the same:
```r
> setkey(two.iris, Sepal.Length, Sepal.Width)
> melt(two.iris)
     Species     variable value
      <fctr>       <fctr> <num>
1:    setosa Sepal.Length   5.1
2: virginica Sepal.Length   5.9
3:    setosa  Sepal.Width   3.5
4: virginica  Sepal.Width   3.0
5:    setosa Petal.Length   1.4
6: virginica Petal.Length   5.1
7:    setosa  Petal.Width   0.2
8: virginica  Petal.Width   1.8
Warning message:
In melt.data.table(two.iris) :
  id.vars and measure.vars are internally guessed when both are 'NULL'. All non-numeric/integer/logical type columns are considered id.vars, which in this case are columns [Species, ...]. Consider providing at least one of 'id' or 'measure' vars in future.
```
Proposal would be to change output above to:
```r
> melt(two.iris, id.vars=c("Sepal.Length", "Sepal.Width"))
   Sepal.Length Sepal.Width     variable     value
          <num>       <num>       <fctr>    <char>
1:          5.1         3.5 Petal.Length       1.4
2:          5.9         3.0 Petal.Length       5.1
3:          5.1         3.5  Petal.Width       0.2
4:          5.9         3.0  Petal.Width       1.8
5:          5.1         3.5      Species    setosa
6:          5.9         3.0      Species virginica
Warning message:
In melt.data.table(two.iris, id.vars = c("Sepal.Length", "Sepal.Width")) :
  'measure.vars' [Petal.Length, Petal.Width, Species, ...] are not all of the same type. By order of hierarchy, the molten data value column will be of type 'character'. All measure variables not of type 'character' will be coerced too. Check DETAILS in ?melt.data.table for more on coercion.
```
It is not clear to me what the benefit would be, but it clearly would be backwards-incompatible, so I suggest closing.

Second proposition is setting a key on output. Currently there is no key. So setting a key would also change the order, potentially backwards incompatible. Closing for now, please feel free to re-open if @pearsonca can find a compelling use case to warrant this backwards-incompatible change.

**Comment by pearsonca** [comment]
This seems an unlikely use case - people might order on Sepal features, but they aren't really keys (though, sure, people use keying as a handy way to accomplish ordering).

I'm thinking more like (with `two.iris` as above):

```r
> setkeyv(two.iris, Species)
> melt(two.iris)
> ... # inferred id.vars, measure.vars exactly as above
> melt(two.iris, id.vars = data.table::key, keyed = TRUE)
Key: <Species, variable>
     Species     variable value
      <fctr>       <fctr> <num>
1:    setosa Sepal.Length   5.1
2:    setosa  Sepal.Width   3.5
3:    setosa Petal.Length   1.4
4:    setosa  Petal.Width   0.2
5: virginica Sepal.Length   5.9
6: virginica  Sepal.Width   3.0
7: virginica Petal.Length   5.1
8: virginica  Petal.Width   1.8
> setkeyv(two.iris, Sepal.Length, Sepal.Width) # the breaking case you describe
> melt(two.iris)
> ... # inferred id.vars, measure.vars exactly as above; requires id.vars to be a function to switch to key-based behavior
```
The new syntax would be:
 - if `id.var` is a function, it's to two.iris to generate id.vars
 - if keyed is truthy:
   * if strictly a logical and true, set the keys to the id.vars (however computed) and the variables (again, however computed)
   * if a non-zero length character vector, set the key to those columns
   * breaking but in my opinion useful: possibly also make keyed have a default behavior of TRUE if set.diff(id.vars, key(data)) is length 0 

introducing the option for id.var to be a function & a new argument with a default (if that default is FALSE) is backwards compatible

I do think my personal preference would be for a change to the null behavior (to check for a key), but that agree that would be breaking. A less breaking (it would occasionally change ordering, but otherwise id/measure vars imputation is identical) change would be to have the `keyed` argument default compute off the data key(s) and id.vars argument.
